Linker error on Windows with chocolate

Here is the log exhibiting the linker failure with cabal 2.2.0.0: https://ci.appveyor.com/project/Mikolaj/allure/build/1.0.148

```
Linking dist\build\Allure\Allure.exe ...
Warning: resolving _GetModuleFileNameW by linking to _GetModuleFileNameW@12
Use --enable-stdcall-fixup to disable these warnings
Use --disable-stdcall-fixup to disable these fixups
    `_GetModuleFileNameW' referenced in section `.text' of dist\build\Allure\Allure-tmp\Paths_Allure.o: defined in discarded section `.text' of C:/ProgramData/chocolatey/lib/ghc/tools/ghc-8.2.2/mingw/bin/../lib/gcc/i686-w64-mingw32/6.2.0/../../../../i686-w64-mingw32/lib/../lib/libkernel32.a(dghgds00535.o)
collect2.exe: error: ld returned 1 exit status
```

And the same project succeeds after downgrading cabal to 2.0.0.1: https://ci.appveyor.com/project/Mikolaj/allure/build/1.0.149

I both cases, initial 64bit build succeed fine, only the subsequent 32bit build fails (with the new cabal version).

I tried `--enable-stdcall-fixup`, which made the warnings disappear, but the error remain. Then I tried `--disable-stdcall-fixup` and I got a different error (as before, only in 32bit build, https://ci.appveyor.com/project/Mikolaj/allure/build/1.0.150):

```
Linking dist\build\Allure\Allure.exe ...
dist\build\Allure\Allure-tmp\Paths_Allure.o:fake:(.text+0xe65): undefined reference to `GetModuleFileNameW'
collect2.exe: error: ld returned 1 exit status
```

When comparing the logs, nothing strikes me, except that only the 64bit compilation (the one that doesn't fail) shows the warning (but only with the old cabal):

```
dist\build\Allure\autogen\Paths_Allure.hs:73:1: warning: [-Wunsupported-calling-conventions]
    * the 'stdcall' calling convention is unsupported on this platform,
      treating as ccall
    * When checking declaration:
        foreign import stdcall unsafe "windows.h GetModuleFileNameW" c_GetModuleFileName
          :: Ptr () -> CWString -> Int32 -> IO Int32
   |
73 | foreign import stdcall unsafe "windows.h GetModuleFileNameW"
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
```

(So it seems as if old cabal had the stdcall problem only on 64bit, but the fixes changed also how 32bit builds work and broke them.)

Also, the log shows some new chocolatey Windows shims are present in the new chocolatey cabal package. But they probably only add the exes to path, so should be benign. 

The error does not appear if the exe is in the same cabal package as the lib: https://ci.appveyor.com/project/Mikolaj/lambdahack/build/1.0.428

Some vaguely related links:

https://mcuoneclipse.com/2017/11/12/solving-problem-with-gnu-linker-and-referenced-in-section-defined-in-discarded-section-error-message
https://stackoverflow.com/questions/31085538/function-referenced-in-section-text-of-defined-in-discarded-section-te
https://ghc.haskell.org/trac/ghc/ticket/9524
https://ghc.haskell.org/trac/ghc/ticket/14300

--- a/Cabal/Distribution/Simple/Build/PathsModule.hs
+++ b/Cabal/Distribution/Simple/Build/PathsModule.hs
@@ -243,13 +243,13 @@ generate pkg_descr lbi clbi =
 
         paths_modulename = autogenPathsModuleName pkg_descr
 
-        get_prefix_stuff = get_prefix_win32 buildArch
+        get_prefix_stuff = get_prefix_win32 supports_cpp buildArch
 
         path_sep = show [pathSeparator]
 
         supports_cpp = supports_language_pragma
-        supports_rebindable_syntax= ghc_newer_than (mkVersion [7,0,1])
         supports_language_pragma = ghc_newer_than (mkVersion [6,6,1])
+        supports_rebindable_syntax= ghc_newer_than (mkVersion [7,0,1])
 
         ghc_newer_than minVersion =
           case compilerCompatVersion GHC (compiler lbi) of
@@ -280,8 +280,8 @@ get_prefix_reloc_stuff =
   "  let (bindir,_) = splitFileName exePath\n"++
   "  return ((bindir `minusFileName` bindirrel) `joinFileName` dirRel)\n"
 
-get_prefix_win32 :: Arch -> String
-get_prefix_win32 arch =
+get_prefix_win32 :: Bool -> Arch -> String
+get_prefix_win32 supports_cpp arch =
   "getPrefixDirRel :: FilePath -> IO FilePath\n"++
   "getPrefixDirRel dirRel = try_size 2048 -- plenty, PATH_MAX is 512 under Win32.\n"++
   "  where\n"++
@@ -295,12 +295,23 @@ get_prefix_win32 arch =
   "              return ((bindir `minusFileName` bindirrel) `joinFileName` dirRel)\n"++
   "            | otherwise  -> try_size (size * 2)\n"++
   "\n"++
+  (case supports_cpp of
+    False -> ""
+    True  -> "#if defined(i386_HOST_ARCH)\n"++
+             "# define WINDOWS_CCONV stdcall\n"++
+             "#elif defined(x86_64_HOST_ARCH)\n"++
+             "# define WINDOWS_CCONV ccall\n"++
+             "#else\n"++
+             "# error Unknown mingw32 arch\n"++
+             "#endif\n")++
   "foreign import " ++ cconv ++ " unsafe \"windows.h GetModuleFileNameW\"\n"++
   "  c_GetModuleFileName :: Ptr () -> CWString -> Int32 -> IO Int32\n"
-    where cconv = case arch of
-                  I386 -> "stdcall"
-                  X86_64 -> "ccall"
-                  _ -> error "win32 supported only with I386, X86_64"
+    where cconv = if supports_cpp
+                     then "WINDOWS_CCONV"
+                     else case arch of
+                            I386 -> "stdcall"
+                            X86_64 -> "ccall"
+                            _ -> error "win32 supported only with I386, X86_64"
 
 filename_stuff :: String
 filename_stuff =

--- a/Cabal/Distribution/Simple/Build/PathsModule.hs
+++ b/Cabal/Distribution/Simple/Build/PathsModule.hs
@@ -248,8 +248,8 @@ generate pkg_descr lbi clbi =
         path_sep = show [pathSeparator]
 
         supports_cpp = supports_language_pragma
-        supports_language_pragma = ghc_newer_than (mkVersion [6,6,1])
         supports_rebindable_syntax= ghc_newer_than (mkVersion [7,0,1])
+        supports_language_pragma = ghc_newer_than (mkVersion [6,6,1])
 
         ghc_newer_than minVersion =
           case compilerCompatVersion GHC (compiler lbi) of

